File: python/devices/memory/ram.py
from numba.experimental import jitclass
from numba import types
from numba import jit
import numpy

class RAM_DICT:
    def __init__(self):
        self.memory = {}
        self.logger = logger

    def read(self, from_addr: int) -> bytes:
        raise NotImplementedError("TODO: implement!")

    def write(self, to_addr: int, data: bytearray):
        raise NotImplementedError("TODO: implement!")

@jitclass([('size', types.int32), ('memory', types.uint8[:])])
class RAM_BYTEARRAY_JIT:
    def __init__(self, size: int):
        self.size = size
        self.memory = numpy.zeros(size, dtype=numpy.uint8)

    def write(self, to_addr: int, data: list[int]):
        if to_addr + len(data) > self.size:
            raise MemoryError("Invalid memory address or size")
        for i in range(len(data)):
            self.memory[to_addr+i] = data[i]

    def read(self, from_addr: int, amount: int) -> list[int]:
        if amount == 1:
            if 0 <= from_addr < self.size:
                return [self.memory[from_addr]]
            else:
                raise MemoryError(f"Invalid memory address: {from_addr}")
        else:
            return [self.memory[from_addr+i] for i in range(amount)]


# Used as an adapter to AddressBus as numba does not know what bytearray is...
class RAM_BYTEARRAY:
    def __init__(self, size: int):
        self.ram = RAM_BYTEARRAY_JIT(size)

    def write(self, to_addr: int, data: bytearray):
        return self.ram.write(to_addr, list(data))

    def read(self, from_addr: int, amount: int):
        return bytearray(self.ram.read(from_addr, amount))

File: lua/devices/memory/ram.lua
local RAM_BYTEARRAY = {}

local function deepcopy(o, seen)
  seen = seen or {}
  if o == nil then return nil end
  if seen[o] then return seen[o] end
  local no = {}
  seen[o] = no
  setmetatable(no, deepcopy(getmetatable(o), seen))
  for k, v in next, o, nil do
    k = (type(k) == 'table') and deepcopy(k, seen) or k
    v = (type(v) == 'table') and deepcopy(v, seen) or v
    no[k] = v
  end
  return no
end

function RAM_BYTEARRAY.init(size)
    local this = deepcopy(RAM_BYTEARRAY)
    this.memory = {}
    this.size = size
    for i=0,size do
        this.memory[i] = 0
    end
    return this
end

function RAM_BYTEARRAY:write(to_addr, data)
    if to_addr + #data > self.size then
        error(string.format("MemoryError: Invalid memory address or size (addr %08x + datasize %08x (%08x) > size %08x)", to_addr, #data, to_addr + #data, self.size))
    end
    for i=1,#data do
        self.memory[to_addr+i-1] = data:byte(i)
    end
end

function RAM_BYTEARRAY:read(from_addr, amount)
    if amount == 1 then
        if from_addr < 0 or from_addr > self.size then
            error("MemoryError: Invalid memory address")
        end
        return string.char(self.memory[from_addr] or 0)
    else
        local result = ""
        for i=from_addr,from_addr+amount-1 do
            if i < 0 or i > self.size then
                error("MemoryError: Invalid memory address")
            end
            result = result .. string.char(self.memory[i] or 0)
        end
        return result
    end
end

return {ram=RAM_BYTEARRAY}